client/resource_group/controller/controller.go

```diff
@@ -429,19 +439,57 @@ func (c *ResourceGroupsController) sendTokenBucketRequests(ctx context.Context,
 // OnRequestWait is used to check whether resource group has enough tokens. It maybe needs to wait some time.
 func (c *ResourceGroupsController) OnRequestWait(
 	ctx context.Context, resourceGroupName string, info RequestInfo,
-) (*rmpb.Consumption, error) {
+) (*rmpb.Consumption, Delta, error) {
 	gc, err := c.tryGetResourceGroup(ctx, resourceGroupName)
 	if err != nil {
 		failedRequestCounter.WithLabelValues(resourceGroupName).Inc()
-		return nil, err
+		return nil, Delta{}, err
+	}
+	consumption, err := gc.onRequestWait(ctx, info)
+	if err != nil {
+		return nil, Delta{}, err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	m, ok := c.resourceDelta[resourceGroupName]
+	if !ok {
+		m = make(map[uint64]Delta)
+		c.resourceDelta[resourceGroupName] = m
 	}
-	return gc.onRequestWait(ctx, info)
+	delta := m[info.StoreID()]
+	// More accurately, it should be reset when the request succeed. But it would cause all concurrent requests piggyback large delta which inflates penalty.
+	// So here resets it directly as failure is rare.
+	m[info.StoreID()] = Delta{}
+
+	return consumption, delta, nil
 }
 
 // OnResponse is used to consume tokens after receiving response
 func (c *ResourceGroupsController) OnResponse(
 	resourceGroupName string, req RequestInfo, resp ResponseInfo,
 ) (*rmpb.Consumption, error) {
+	// record resource delta
+	c.mutex.Lock()
+	if resp.Succeed() {
+		m, ok := c.resourceDelta[resourceGroupName]
+		if !ok {
+			m = make(map[uint64]Delta)
+			c.resourceDelta[resourceGroupName] = m
+		}
+		for id, delta := range m {
```

Comment:
There is no need to increase the delta for all groups, but instead you can just maintain a global non-decrease count and add the delta value after each request. And in `OnRequestWait`, the delta can be calculate by `globalCounter - storeCounter` and then set storeCounter to globalCounter, that's it.